monitor/cert_submitter.go

```diff
@@ -60,63 +98,91 @@ type certSubmitter struct {
 			c.clk.Sleep(c.certSubmitInterval)
 		}
 	}()
 }
 
-// submitCertificate issues a certificate with the certSubmitter's
-// certIssuer/certIssuerKey and submits it to a monitored log's add-chain
-// endpoint. The latency of the submission is tracked in the
-// `cert_submit_latency` prometheus histogram. If the submission fails, or the
-// returned SCT is invalid the `cert_submit_results` prometheus countervec is
-// incremented with a "fail" status tag. If the submission succeeds the
-// `cert_submit_results` prometheus countervec is incremented with a "ok" status
-// tag. An SCT is considered invalid if the signature does not validate, or if
-// the timestamp is too far in the future or the past (controlled by
-// `requiredSCTFreshness`).
-func (c *certSubmitter) submitCertificate() {
+// submitCertificates issues a pre-certificate and a matching certificate with
+// the certSubmitter's certIssuer/certIssuerKey. If `submitPreCert` is enabled
+// then a precert is submitted. If `submitCert` is enabled then a final cert is
+// submitted.
+func (c *certSubmitter) submitCertificates() {
 	// a certSubmitter requires a non-nil certIssuerKey and certIssuer. If somehow
 	// was one created with nil values then panic.
 	if c.certIssuerKey == nil || c.certIssuer == nil {
 		panic("certSubmitter created with nil certIssuerKey or certIssuer\n")
 	}
 
-	c.logger.Printf("Submitting certificate to %q\n", c.logURI)
-
-	cert, err := pki.IssueTestCertificate(c.certIssuerKey, c.certIssuer, c.clk)
+	certPair, err := pki.IssueTestCertificate(c.certIssuerKey, c.certIssuer, c.clk)
 	if err != nil {
 		// This should not occur and if it does we should abort hard
 		panic(fmt.Sprintf("!!! Error issuing certificate: %s\n", err.Error()))
 	}
 
-	// Because ct-woodpecker issues directly off of a fake root the "chain" only
-	// contains the leaf certificate we minted in issueCertificate()
+	if c.submitPreCert {
+		c.submitCertificate(certPair.PreCert, true)
+	}
+
+	if c.submitCert {
+		c.submitCertificate(certPair.Cert, false)
+	}
+}
+
+// submitCertificate submits a single x509 Certificate to a log. If `isPreCert`
+// is true then the certificate is submitted via the log's add-pre-chain
+// endpoint, otherwise the add-chain endpoint is used. The latency of the
+// submission is tracked in the `cert_submit_latency` prometheus histogram. If
+// the submission fails, or the returned SCT is invalid the
+// `cert_submit_results` prometheus countervec is incremented with a "fail"
+// status tag. If the submission succeeds the `cert_submit_results` prometheus
+// countervec is incremented with a "ok" status tag. An SCT is considered
+// invalid if the signature does not validate, or if the timestamp is too far in
+// the future or the past (controlled by `requiredSCTFreshness`).
+func (c certSubmitter) submitCertificate(cert *x509.Certificate, isPreCert bool) {
 	chain := []ct.ASN1Cert{
 		{Data: cert.Raw},
 	}
 
+	// Precert submissions also need the issuer in the chain
```

Comment:
The SCT struct for a `PrecertChainEntry` contains the issuer SPKI whereas for `X509ChainEntry ` it does not. The trillian CT frontend requires that this come from a issuer included in the add-pre-chain request. In theory it could build the chain itself and add it from the root itself if no issuer was passed in, but that's likely only to ever affect testing cases as no one will be issuing directly from a root.

Reply:
@rolandshoemaker Thanks for the explanation :+1: I added a little bit more detail to the comment here based on your words.

@jsha to put it into concrete terms the `cttestsrv` uses `ct.MerkleTreeLeafFromRawChain` to construct the SCT it returns: https://github.com/letsencrypt/ct-woodpecker/blob/046129d977ce3ec8ffc8f6ba4ed904a6dbe65a0e/test/cttestsrv/sct.go#L33 

That in turn ends up calling `ct.MerkleTreeLeafFromChain` which bails immediate if the provided chain length is less than 2: https://github.com/google/certificate-transparency-go/blob/76406a93aa88c8ce3b42c8c1a74ae4197b9c59ab/serialization.go#L157-L161

As Roland mentioned, we could _probably_ submit a patch to the upstream but it seems like more work than its worth given that we're only tickling this in a test context issuing directly off of a root and we can work around it trivially by just sending the issuer in the chain.